src/BlockTypes/ProductQuery.php

```diff
@@ -79,6 +79,60 @@ protected function initialize() {
 		);
 		add_filter( 'rest_product_query', array( $this, 'update_rest_query' ), 10, 2 );
 		add_filter( 'rest_product_collection_params', array( $this, 'extend_rest_query_allowed_params' ), 10, 1 );
+		add_filter( 'render_block_core/query', array( $this, 'add_navigation_id_directive' ), 10, 3 );
+		add_filter( 'render_block_core/query-pagination', array( $this, 'add_navigation_link_directives' ), 10, 3 );
+	}
+
+	/**
+	 * Mark the Product Query as an interactive region so it can be updated
+	 * during client-side navigation.
+	 *
+	 * @param string    $block_content The block content.
+	 * @param array     $block         The full block, including name and attributes.
+	 * @param \WP_Block $instance      The block instance.
+	 */
+	public function add_navigation_id_directive( $block_content, $block, $instance ) {
+		if ( self::is_woocommerce_variation( $block ) ) {
+			// Enqueue the Interactivity API runtime.
+			wp_enqueue_script( 'wc-interactivity' );
+
+			$p = new \WP_HTML_Tag_Processor( $block_content );
+
+			// Add `data-wc-navigation-id to the query block.
+			if ( $p->next_tag( array( 'class_name' => 'wp-block-query' ) ) ) {
+				$p->set_attribute( 'data-wc-interactive', true );
```

Comment:
There's not a real need to hydrate the inner content of the Products block yet, is it? Maybe we can just hydrate the pagination for now…

Reply:
I'll try and see if it works as expected (I guess so). 👍

Reply:
It works; fixed in https://github.com/woocommerce/woocommerce-blocks/pull/10200/commits/7b4c991f5d735c511c9d75e3a29be27a96684715 ✅

Reply:
The issue was that Preact was instantiating two different root fragments, one for the Query block and another for the Query Pagination block, thus leading to inconsistencies and issues reconciling the HTML.

Additionally, root fragments were instantiated for every client-side navigation instead of reusing them.

Reply:
Thanks for investigating that possibility, David.